security/access_control.rst

```diff
@@ -288,13 +305,26 @@ key:
             'access_control' => [
                 [
                     'path' => '^/_internal/secure',
-                    'allow_if' => '"127.0.0.1" == request.getClientIp() or has_role("ROLE_ADMIN")',
+                    'roles' => 'ROLE_ADMIN',
+                    'allow_if' => '"127.0.0.1" == request.getClientIp() or request.header.has('X-Secure-Access')',
                 ],
             ],
 
-In this case, when the user tries to access any URL starting with ``/_internal/secure``,
-they will only be granted access if the IP address is ``127.0.0.1`` or if
-the user has the ``ROLE_ADMIN`` role.
+In this case, when the user tries to access any URL starting with
+``/_internal/secure``, they will only be granted access if the IP address is
+``127.0.0.1`` or a secure header, or if the user has the ``ROLE_ADMIN`` role.
+
+.. note::
+
+    Internally ``allow_if`` enforce calling the built-in
+    :class:`Symfony\\Component\\Security\\Core\\Authorization\\Voter\\ExpressionVoter`
+    as like it was part of the attributes defined in the ``roles`` option.
```

Comment:
"as like it was part" - I don't understand this part. Is the `$attributes` passed to this voter == to the `allow_if` string? Is that what we're describing?

Reply:
Yeah we need to convert the string to an expression so it is evaluated by the `ExpressionVoter`, basically it should have the same effect as passing the string to the `roles` array directly.
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Core/Authorization/Voter/ExpressionVoter.php#L76. This is also what handles the `@Security` annotation to evaluate its value.

Reply:
"should", I mean it does not actually. And we should keep it that way IMHO.

Reply:
Ok re-reading your comment, it acts as I say, but I have to answer "no" to what you thought it was.

What I want to explain is how `isGranted($attributes, $subject)` is called, the following is happening:

```yaml
// config/packages/security.yaml
security:
    # ...
    access_control:
        - { path: /, roles: [ROLE_USER, ROLE_ADMIN], condition: 'request.isMethod("GET")' }
```
will be equivalent to:
```php
$authCheck->isGranted([
    'ROLE_USER',
    'ROLE_ADMIN',
    new Expression('request.isMethodd("GET")'),
], $request);
```
Each voter get all attributes and the request as subject, the token (and any user) by contract, with the default strategy one voter granting one attribute allow access, even is there was some denying.